File: lib/services/GuestBookService.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/material.dart';
import '../models/StickyNoteModel.dart';

class GuestBookService {
  final _db = FirebaseFirestore.instance;

  CollectionReference<Map<String, dynamic>> _col(String targetUserId) =>
      _db.collection('users').doc(targetUserId).collection('guestbook');

  Stream<List<StickyNoteModel>> watchNotes(String targetUserId) {
    return _col(targetUserId)
        .orderBy('pinned', descending: true)
        .orderBy('createdAt', descending: true)
        .snapshots()
        .map((s) => s.docs.map(_fromDoc).toList());
  }

  Future<void> addNote(String targetUserId, StickyNoteModel note) async {
  final ref = _col(targetUserId).doc(note.id);
    await ref.set({
      'id': note.id,
      'authorId': note.authorId,
      'authorName': note.authorName,
      'authorAvatarUrl': note.authorAvatarUrl,
      'text': note.text,
      'color': note.color.value,
      'pinned': note.pinned,
      'rotationDeg': note.rotationDeg, // 여기는 그대로 note에서 받아서 저장
      'createdAt': FieldValue.serverTimestamp(),
    }, SetOptions(merge: true));
  }

  Future<void> updateNote(String targetUserId, StickyNoteModel note) async {
    await _col(targetUserId).doc(note.id).update({
      'text': note.text,
      'color': note.color.value,
      'rotationDeg': note.rotationDeg,
      'pinned': note.pinned,
    });
  }

  Future<void> deleteNote(String targetUserId, String noteId) async {
    await _col(targetUserId).doc(noteId).delete();
  }

  StickyNoteModel _fromDoc(QueryDocumentSnapshot<Map<String, dynamic>> d) {
    final m = d.data();
    return StickyNoteModel(
      id: m['id'] as String,
      authorId: m['authorId'] as String,
      authorName: (m['authorName'] ?? '') as String,
      authorAvatarUrl: (m['authorAvatarUrl'] ?? '') as String,
      createdAt: (m['createdAt'] as Timestamp?)?.toDate() ?? DateTime.now(),
      text: (m['text'] ?? '') as String,
      color: Color(m['color'] as int),
      pinned: (m['pinned'] ?? false) as bool,
      rotationDeg: (m['rotationDeg'] ?? 0).toDouble(), // 저장된 값 그대로 사용
    );
  }

}

File: ting-rn/src/services/guestBookService.ts
import firestore from '@react-native-firebase/firestore';
import { StickyNote } from '../types/guestbook';

function col(targetUserId: string) {
  return firestore()
    .collection('users')
    .doc(targetUserId)
    .collection('guestbook');
}

function fromDoc(d: any): StickyNote | null {
  const m = d?.data?.();
  if (!m) return null;
  return {
    id: (m.id as string) ?? d.id,
    authorId: (m.authorId as string) ?? '',
    authorName: (m.authorName as string) ?? '',
    authorAvatarUrl: (m.authorAvatarUrl as string) ?? '',
    createdAt: m.createdAt?.toDate?.() ?? new Date(),
    text: (m.text as string) ?? '',
    color: (m.color as number) ?? 0,
    pinned: (m.pinned as boolean) ?? false,
  };
}

export const guestBookService = {
  /** Returns a query for real-time listener */
  watchQuery(targetUserId: string) {
    return col(targetUserId)
      .orderBy('pinned', 'desc')
      .orderBy('createdAt', 'desc');
  },

  async addNote(targetUserId: string, note: StickyNote): Promise<void> {
    await col(targetUserId).doc(note.id).set(
      {
        id: note.id,
        authorId: note.authorId,
        authorName: note.authorName,
        authorAvatarUrl: note.authorAvatarUrl,
        text: note.text,
        color: note.color,
        pinned: note.pinned,
        createdAt: firestore.FieldValue.serverTimestamp(),
      },
      { merge: true },
    );
  },

  async updateNote(targetUserId: string, note: StickyNote): Promise<void> {
    await col(targetUserId).doc(note.id).update({
      text: note.text,
      color: note.color,
      pinned: note.pinned,
    });
  },

  async deleteNote(targetUserId: string, noteId: string): Promise<void> {
    await col(targetUserId).doc(noteId).delete();
  },

  parseDoc: fromDoc,
};
